### ISSUE TITLE
How to form queries with time filters?

### ISSUE THREAD

**Issue by mkarlesky** [Synergetic-Engineering/odata-influxdb#17]
We're back. Our setup is working fairly well (thanks again), but we've hit a stumbling block.

As InfluxDB is a time-series database we'd like to execute time-based queries. So far we've had no luck. Of course, we're new to OData and there are limitations to the present state of this project, so it's entirely possible we're simply doing it wrong.

With the following URL that filters by ‘Location’, we get entries for that location that look like below:

> `http://odata.acme.com:8080/sensors__Water__sp__Temperature?$top=10&$filter=Location%20eq%20%27South%20Tank%27`

```
<entry>
  <id>
    http://odata.acme.com:8080/sensors__Water__sp__Temperature(datetime'2017-01-12T01%3A00%3A24')
  </id>
  <title type="text"/>
  <updated>2017-05-26T19:43:33Z</updated>
  <link href="http://odata.acme.com:8080/sensors__Water__sp__Temperature(datetime'2017-01-12T01%3A00%3A24')" rel="edit"/>
  <category scheme="http://schemas.microsoft.com/ado/2007/08/dataservices/scheme" term="InfluxDBSchema.sensors__Water__sp__Temperature"/>
  <content type="application/xml">
    <m:properties>
      <d:time>2017-01-12T01:00:24</d:time>
      <d:Location>South Tank</d:Location>
      <d:Value>27.67</d:Value>
      <d:Units>°C</d:Units>
    </m:properties>
  </content>
</entry>
```

This looks right and makes sense.

However, if we change the query to specifically reference the timestamp in the above entry, we see an error.

> `http://odata.acme.com:8080/sensors__Water__sp__Temperature?$top=10&$filter=time%20eq%20datetime%272017-01-12T01%3A24%3A35%27`

```
<error xmlns="http://schemas.microsoft.com/ado/2007/08/dataservices/metadata">
  <code>InvalidSystemQueryOption</code>
  <message>
    Invalid System Query Option: $filter : ParserError: expected time literal at [4]
  </message>
</error>
```

Questions:

1. What is the proper way to specify a datetime literal in a query? According to the [OData v2 spec](http://www.odata.org/documentation/odata-version-2-0/overview/), we think we're doing it correctly -- ex. `datetime'2000-12-12T12:00'`.
1. Is it possible there's a mismatch between odata-influxdb and our version of InfluxDB (1.1)?
1. Is the project missing support among its implemented filter operators (gt, lt, eq, ne) for time values? If so, how might we go about adding that?

Bonus Question:

- Using odata-influxdb as currently implemented, is it possible to construct a generic query URI that returns data for the most recent _x_ timespan without having to embed a reference date (e.g. Dear OData source, please give me all the data for the most recent two weeks in your datastore.)?

**Comment by srp-synengco** [comment]
Thanks for this, @soundstripe has been working on a pull request for this type of thing, so I'll leave it to him to comment on how he's going about that

and yeah, good point for the bonus question, I'd have to do a bit more reading about how to best handle that in odata  - in general, the broader question of 'what best-practices there are for working with time-series data through odata' is something that I should look into more thoroughly, as i have found that quering a lot of data can be slow (since it constructs and downloads all the metadata for each row)

**Comment by soundstripe** [comment]
There's a few example queries in the README now.

The queries were designed according to InfluxDB 1.2 specs, but I believe they will work for 1.1 with no issues.

I have added to the supported filter operators, and the code to add more is pretty simple if you find some you need to be missing.

As for the Bonus Question, there are query functions in the OData spec (such as "now()") but it doesn't look like pyslet implements these. I think it would be possible to put this in this project, but would be better as a contribution to pyslet (toward its existing issue to support OData v4 etc).

**Comment by mkarlesky** [comment]
Thanks for the quick response! I can report that the latest version does allow us to successfully run time-based queries on our data, and we get entirely sensible results. Also tried a compound filter with timestamp AND 'Location', and it worked perfectly as well.

Towards answering our own Bonus Question… Hacking odata-influxdb to support `now()` feels yucky. And I don't expect pyslet will have stable support for OData v4 anytime soon. I think in our specific application we can use Nginx or something similar to wire up URL handling such that we can effectively insert/expand `now()` in a query URL. That should get us where we want to go fairly painlessly.

**Comment by soundstripe** [comment]
Closing after fixes from #18
